nodejs/src/lib/inference-session-impl.ts

```diff
@@ -0,0 +1,222 @@
+import {binding, Binding} from './binding';
+import {InferenceSession as InferenceSessionInterface, InferenceSessionFactory} from './inference-session';
+import {OnnxValue} from './onnx-value';
+import {Tensor} from './tensor';
+
+type SessionOptions = InferenceSessionInterface.SessionOptions;
+type RunOptions = InferenceSessionInterface.RunOptions;
+type FeedsType = InferenceSessionInterface.FeedsType;
+type FetchesType = InferenceSessionInterface.FetchesType;
+type ReturnType = InferenceSessionInterface.ReturnType;
+
+class InferenceSession implements InferenceSessionInterface {
+  constructor(session: Binding.InferenceSession) {
+    this.#session = session;
+    // cache metadata
+    this.inputNames = this.#session.inputNames;
+    this.outputNames = this.#session.outputNames;
+  }
+  run(feeds: FeedsType, options?: RunOptions): Promise<ReturnType>;
+  run(feeds: FeedsType, fetches: FetchesType, options?: RunOptions): Promise<ReturnType>;
+  run(feeds: FeedsType, arg1?: FetchesType|RunOptions, arg2?: RunOptions): Promise<ReturnType> {
+    const fetches: {[name: string]: OnnxValue|null} = {};
+    let options: RunOptions = {};
+    // check inputs
+    if (typeof feeds !== 'object' || feeds === null || feeds instanceof Tensor || Array.isArray(feeds)) {
+      throw new TypeError(
+          '\'feeds\' must be an object that use input names as keys and OnnxValue as corresponding values.');
+    }
+
+    let isFetchesEmpty = true;
+    // determine which override is being used
+    if (typeof arg1 === 'object') {
+      if (arg1 === null) {
+        throw new TypeError('Unexpected argument[1]: cannot be null.');
+      }
+      if (arg1 instanceof Tensor) {
+        throw new TypeError('\'fetches\' cannot be a Tensor');
+      }
+
+      if (Array.isArray(arg1)) {
+        if (arg1.length === 0) {
+          throw new TypeError('\'fetches\' cannot be an empty array.');
+        }
+        isFetchesEmpty = false;
+        // output names
+        for (const name of arg1) {
+          if (typeof name !== 'string') {
+            throw new TypeError('\'fetches\' must be a string array or an object.');
+          }
+          if (this.outputNames.indexOf(name) === -1) {
+            throw new RangeError(`'fetches' contains invalid output name: ${name}.`);
+          }
+          fetches[name] = null;
+        }
+
+        if (typeof arg2 === 'object') {
+          options = arg2;
+        } else if (typeof arg2 !== 'undefined') {
```

Comment:
Since `arg2` is a declared variable, `if(arg2)` is sufficient. Up to you, `typeof a !== 'undefined'` is typically when `a` may not be defined.

Reply:
`if (arg2)` has different behavior to `if (typeof arg2 !== 'undefined')`.

this code snippet is checking the type of each argument and try to catch every type mismatch error.

`arg2 !== undefined` should be equivalent to `typeof arg2 !== 'undefined'`. But as the reason explained above, and considering JS engine has done as much as we can imagine to optimize the JIT there is not much difference on performance. Simply stick to `typeof arg2 !== 'undefined'` should never be wrong.